add homework 03

aspanfilov opened the pull request:
add homework 03

aspanfilov committed:
add homework03 core

aspanfilov committed:
add HomeWork 03

petrelevich commented on hw03-reflection/src/main/java/ru/otus/WorkingClass.java:
```diff
@@ -0,0 +1,4 @@
+package ru.otus;
+
+public class WorkingClass {
```
зачем этот класс?

aspanfilov replied:
Это, предполагалось, что заглушка для некого рабочего класса, методы которого тестируем в классе WorkingClassTest

petrelevich commented on hw03-reflection/src/main/java/ru/otus/test/Annotation/After.java:
```diff
@@ -0,0 +1,11 @@
+package ru.otus.test.Annotation;
```
в названии пакетов не принято использовать большие буквы.

aspanfilov replied:
принято

petrelevich commented on hw03-reflection/src/main/java/ru/otus/test/unit/WorkingClassTest.java:
```diff
@@ -0,0 +1,36 @@
+    public void methodTest1() throws Exception {
+        Exception e = new Exception("Ошибка ввода вывода");
+        throw e;
+//        System.out.println("run method Test 1");
```
закоментированный код не надо сохранять в гит.

aspanfilov replied:
Для этого есть какая-то настройка для .gitignore ? 
В данном случае закомментирована отображалка успешного прохождения теста, чтобы можно было ее раскомментировать, для экспериментов. И добавлена какая-то ошибка, чтобы посмотреть как тестирование отработает с ошибкой в методе.

petrelevich replied:
> Для этого есть какая-то настройка для .gitignore ?

Такой настройки нет.

petrelevich commented on hw03-reflection/src/main/java/ru/otus/test/TestStarter.java:
```diff
@@ -0,0 +1,10 @@
+import java.lang.reflect.InvocationTargetException;
+
+public class TestStarter {
+    public static void main(String[] args) throws ClassNotFoundException, InvocationTargetException, NoSuchMethodException, InstantiationException, IllegalAccessException {
```
такой длинный хвост исключений можно заменить на Exception

aspanfilov replied:
принято

petrelevich commented on hw03-reflection/src/main/java/ru/otus/test/TestExecutor.java:
```diff
@@ -0,0 +1,75 @@
+import java.util.Set;
+
+public class TestExecutor {
+    private Set<Method> beforeMethods;
```
поля стоит сделать final

aspanfilov replied:
принято

petrelevich commented on hw03-reflection/src/main/java/ru/otus/test/TestExecutor.java:
```diff
@@ -0,0 +1,75 @@
+    private void executeMethods(Class<?> clazz) throws IllegalAccessException, InvocationTargetException, InstantiationException, NoSuchMethodException {
+
+        Constructor<?>[] constructors = clazz.getConstructors();
+        Object object = constructors[0].newInstance();
```
надежнее брать не первый попавшийся конструктор, а нужный, т.е. конструктор без параметров.

petrelevich replied:
Обратите внимание на важный пункт ДЗ:
> для каждой такой "тройки" надо создать СВОЙ объект класса-теста.

таким образом обеспечивается независимость тестов.
в текущей реализации все тесты используют один объект - object.

aspanfilov replied:
принято

petrelevich commented on hw03-reflection/src/main/java/ru/otus/test/MethodTester.java:
```diff
@@ -0,0 +1,50 @@
+            PrintWriter pw = new PrintWriter(sw);
+            e.printStackTrace(pw);
+
+            return method.getName() + ": \n" + sw.toString();
```
очень плохая практика смешивать выполнение кода и форматирование результата.
Нарушается принцип единства ответственности.
Если надо вернуть результат, то он должен быть без форматирования.

aspanfilov replied:
принято

petrelevich commented on hw03-reflection/src/main/java/ru/otus/test/TestExecutor.java:
```diff
@@ -0,0 +1,75 @@
+        Object object = constructors[0].newInstance();
+
+        for (Map.Entry<Method, String> testMethod : this.testMethods.entrySet()) {
+            MethodTester methodTester = new MethodTester(object, testMethod.getKey(), this.beforeMethods, this.afterMethods);
```
зачем для запуска теста сделан отдельный класс?
почему недостаточно просто метода?

aspanfilov replied:
Это я так реализовал вот эту фразу из дз "для каждой такой "тройки" надо создать СВОЙ объект класса-теста." Причем изначально у меня и был метод, который я затем переделал в класс. 
В-общем, теперь я понял смысл - переделаем.

aspanfilov committed:
add HomeWork 03
Удалил неиспользуемый класс WorkingClass
Поправил нейминг пакета
Удалил комментарии в коде
Сгруппировал перечень исключений в сигнатурах методов
сделал поля final для класса TestExecutor
Отредактировал подбор конструктора рефлексией
Провел изолирование формирование лога
Удалил класс запускалку метода - оформил логику в методе класса TestExecutor
Исправил формирование инстанса тестируемого объекта - на каждый тестируемый метод.
